### PR TITLE
added gamma cdf functions

### LINKED ISSUE #24: Add CDF for Gamma Distribution
[Here](https://github.com/hot9cups/probdists/blob/main/probdists/Gammadistribution.py) is the Gamma distribution class.

It contains the PDF function but currently does not have a CDF function.
I would like it contain the CDF too.

I would also like appropriate unittests to be included in the [test.py file](https://github.com/hot9cups/probdists/blob/main/test.py)

### PR DESCRIPTION
**Pull request by misterdabu**
I have added upper and lower CDF functions to the gammadistribution.py file. I have also added a couple of test cases to the test file to verify each of the functions. This resolves #24 

Please let me know if you have any questions or if any updates are necessary.

Thanks you.

### CODE REVIEW

File: probdists/Gammadistribution.py [hunk 1]
```diff
@@ -102,6 +102,34 @@ def plot_bar_pdf(self, points=25):
...
+                x (float): point for calculating the cumulative distribution function
+            Returns:
+                float: lower cumulative distribution function output (1-cdf)
+        """
+        #initiate cdf variable
+        cdf = 0
+        for i in range (self.k):
+            cdf += (math.pow((x / self.theta), i) * math.exp(-1 * x / self.theta)) / math.factorial(i) 
+        return 1-cdf
+
```
Comment (hot9cups) [inline probdists/Gammadistribution.py:118]: Well correct me if I'm wrong, but isn't this only applicable when k > 0? 
What do we do when k <= 0?

File: probdists/Gammadistribution.py [hunk 2]
```diff
@@ -102,6 +102,34 @@ def plot_bar_pdf(self, points=25):
...
+                x (float): point for calculating the cumulative distribution function
+            Returns:
+                float: upper cumulative distribution function output (cdf)
+        """
+        #initiate cdf variable
+        cdf = 0
+        for i in range (self.k):
+            cdf += (math.pow((x / self.theta), i) * math.exp(-1 * x / self.theta)) / math.factorial(i) 
+        return cdf
+
```
Comment (hot9cups) [inline probdists/Gammadistribution.py:132]: Could you provide a reference for this? I wasn't able to find much information on an upper cdf for gamma distribution on wikipedia.

File: test.py [hunk 3]
```diff
@@ -176,6 +176,14 @@ def test_pdf(self):
...
                          'pdf function does not give expected result')
 
+    def test_cdf_lower(self):
+        self.assertEqual(round(self.gamma.cdf_lower(4), 5), .59399, 
+                         'cdf function does not give expected result')
+
+    def test_cdf_upper(self):
+        self.assertEqual(round(self.gamma.cdf_upper(4), 5), .40601, 
+                         'cdf function does not give expected result')
+
```
Comment (hot9cups) [inline test.py:186]: Ah, @mhdzumair can you review this and check if the round() here adheres to our code style?

**Review by hot9cups — COMMENTED** [review]
@misterdabu Glad to see you again. Thanks for the PR, this looks good, just want to confirm about a couple of things before merging and I've left code review comments in the code for that :)

**Comment by hot9cups** [comment]
There's also some trailing whitespaces in the code that you ought to remove, so that the Codacy tests pass too.

File: probdists/Gammadistribution.py (hunk 2 above)
Comment (mhdzumair) [reply probdists/Gammadistribution.py:132]: i think this isn't a good way to do get cdf value. You wrote the same code twice their only change is
 
in cdf_lower returned  1 - cdf 
and cdf_upper returend the cdf

you can make an argument to set if user asking the lowe or upper. So the user can get cdf with one function. And also you reduce the redundancy.

File: test.py (hunk 3 above)
Comment (mhdzumair) [reply test.py:186]: See other distribution cdf function which is implemented with internal round arg.

**Review by mhdzumair — CHANGES_REQUESTED** [review]

**Comment by misterdabu** [comment]
Thank you very much for the feedback guys! Per @mhdzumair I have updated the function to reflect only 1 function which takes in an "upper" or "lower" argument. 

I got this particular function from [https://en.wikipedia.org/wiki/Gamma_distribution](https://en.wikipedia.org/wiki/Gamma_distribution) for the generalized gamma function which is only valid when x >= 0. I have added validation to make sure the x values are greater than 0 and the gamma class also has validation for theta >= 0. I know the gamma function can handle negative non-integer negative numbers, however, I have not seen much information on how to calculate CDFs for x < 0 programmatically. As such, I excluded negatives from this function lest I make a mistake in my calculation.

Additionally, my submission is giving me a **TypeError: 'NoneType' object is not callable**. I'm not sure what I'm doing wrong here as the function does pass the checks locally without raising an exception. Any help on this while I continue to research the issue would be greatly appreciated.

Thanks you very much again for letting me work on this.

**Review by mhdzumair — CHANGES_REQUESTED** [review]
cdf is an attribute in Distribution class. you cannot use that name for function. thats why its give error. change your function name to calculate_cdf() and also add round_to arg as 3rd argument for your method. then change the result varibale name to self.cdf. then return the self.cdf value with rounded with round_to . for the reference check the calculate_pdf()

**Comment by mhdzumair** [comment]
And one more thing. you can use boolen value for limitType. 
- upper value as True.
- lower value as False.

if you use boolen then change the argument name. use a proper name for it. like `is_upper` then, you can set a default value to True. \
in this way you dont need the exception handle. i think this is much more efficient.

**Comment by misterdabu** [comment]
@mhdzumair, thank you so much for such detailed feedback. I have made all the aforementioned updates in my most recent commit. The round_to variable was a great idea as well. In the test file, I left the upper test with all default values and passed in x=4 and in the lower I defined values within the function call to ensure that they were working correctly. These seems to pass the checks so hopefully, it is ready to be merged. 

Please let me know if further updates are necessary.

**Review by mhdzumair — APPROVED** [review]
yeah, this is fine. could you add the usage of this function in USAGE.md other than that it's alright. ill approve this PR.

**Comment by misterdabu** [comment]
Thanks @mhdzumair. I have updated USAGE.md to reflect the gamma cdf function. Please feel free to reach out if you have any questions or if further updates are necessary.

**Comment by mhdzumair** [comment]
Thanks for contributing this repo @misterdabu. I think now the time to merge this @hot9cups

**Comment by hot9cups** [comment]
Okay this is great!
Thank you @misterdabu for contributing to probdists! I'll be sure to include you as a contributor when the next version gets released(very soon)!
This was good code! 
Oh also, if you find the library helpful and/or know someone who might, do share it! 
Consider giving probdists a star if you like it :)
Thanks again, for the time and effort you put into this !!

Thanks zumair for taking care of this! Agree with all the changes you requested! Code quality is great now!

**Pull request merged**